### PR TITLE
Add Map and Program

### LINKED ISSUE #1: Upstream Map and Program from newtools/ebpf
Pull in the `Map` and `Program` implementation from github.com/newtools/ebpf.

### PR DESCRIPTION
**Pull request by lmb**
Pull in `newtools/ebpf` Map and Program, and preserve history by creating a merge commit.

This includes the `asm` subpackage to be able to test `Program` and `Map` without requiring an ELF loader.

I made the following changes from `newtools/ebpf` master, based on our previous discussion:

```
map: add test for PerfEventArray
program: add test for Benchmark
asm: rename XAdd to StoreXAdd
asm: update and rename built-in functions
map: add benchmarks
readme: update resources section
program: rename receiver
map: remove Marshaler in favour of documentation
doc: make package documentation more concise
abi: improve documentation
program: expose IsNotSupported instead of ErrNotSupported
map: expose Update and remove Create and Replace
map: return a NotExist error from NextKey
map: return error from Delete if key doesn't exist
map: rename Get to Lookup and return not found error
types: rename ProgType to ProgramType
remove helper flags
remove ELF loader
remove perf reader
remove non-self contained examples
remove binaries
remove GitHub templates
```

Fixes #1

### CODE REVIEW

File: license.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Copyright (c) 2017 Nathan Sweet
```
Comment (lmb) [inline license.md]: We'll have to merge this with `LICENSE` in a follow up commit.

File: readme.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+eBPF
```
Comment (brb) [inline readme.md]: Could we add a huge banner saying "Work in progress. The API will definitely change"?

Comment (lmb) [reply readme.md]: Yeah, I'll do that in a PR straight after this if that is OK?

**Comment by lmb** [comment]
@brb @joestringer can either of you approve the PR?

File: readme.md (hunk 2 above)
Comment (brb) [reply readme.md]: Sure.

**Pull request merged**